src/mrpro/data/SpatialDimension.py

```diff
@@ -24,80 +46,401 @@ class XYZ(Protocol[T]):
 
 
 @dataclass(slots=True)
-class SpatialDimension(MoveDataMixin, Generic[T]):
+class SpatialDimension(MoveDataMixin, Generic[T_co]):
     """Spatial dataclass of float/int/tensors (z, y, x)."""
 
-    z: T
-    y: T
-    x: T
+    z: T_co
+    y: T_co
+    x: T_co
 
     @classmethod
-    def from_xyz(cls, data: XYZ[T], conversion: Callable[[T], T] | None = None) -> SpatialDimension[T]:
+    def from_xyz(cls, data: XYZ[T_co]) -> SpatialDimension[T_co]:
         """Create a SpatialDimension from something with (.x .y .z) parameters.
 
         Parameters
         ----------
         data
             should implement .x .y .z. For example ismrmrd's matrixSizeType.
-        conversion,  optional
-            will be called for each value to convert it
         """
-        if conversion is not None:
-            return cls(conversion(data.z), conversion(data.y), conversion(data.x))
         return cls(data.z, data.y, data.x)
 
     @staticmethod
     def from_array_xyz(
         data: ArrayLike,
-        conversion: Callable[[torch.Tensor], torch.Tensor] | None = None,
     ) -> SpatialDimension[torch.Tensor]:
         """Create a SpatialDimension from an arraylike interface.
 
         Parameters
         ----------
         data
             shape (..., 3) in the order (x,y,z)
-        conversion
-            will be called for each value to convert it
         """
-        if not isinstance(data, np.ndarray | torch.Tensor):
-            data = np.asarray(data)
-
+        data_ = _as_vectortype(data)
         if np.size(data, -1) != 3:
```

Comment:
I would have this check before _as_vectortype

Reply:
It should be this order.

ArrayLike is anything that numpy can convert to an Array, for example a list of lists.
Np.shape on objects that do not have .shape is plemented by converting to an Array, then checking it's size.
So switching the commands here would result in double conversion and has no advantages?

Reply:
should it then be np.size(data_, -1)?

Reply:
YES! 
thank you